File: src/dpm/publish.py
import json
import os
import sys
import shutil
from zipfile import ZipFile
from util import *
# TURN OFF YOUR FORMATTER, IT WILL PUT THIS IN THE WRONG ORDER AND IT WILL STOP WORKING!!!
import sys
# sys.path.append(f"C:/Users/{os.getlogin()}/OneDrive/Desktop/Pyrinth/src") # Testing - Windows
sys.path.append(f"../") # Testing - Other OS's
from pyrinth.modrinth import *
from pyrinth.models import *
from pyrinth.users import *
from pyrinth.projects import *

def create_project(namespace):
    namespace = to_slug(namespace)
    name = to_sentence_case(namespace)
    description = input(f"Enter the description of datapack '{name}': ")
    if len(description) < 3:
        print(f"Warning! You will not be able to publish '{name}' to modrinth if the description is under 3 characters!")
    if name == '':
        print("Datapack name cannot be blank")
        return
    tags_functions = f"{name}/data/minecraft/tags/functions"
    functions = f"{name}/data/{namespace}/functions"
    if not os.path.exists(tags_functions):
        os.makedirs(tags_functions)

    if not os.path.exists(functions):
        os.makedirs(functions)

    if not os.path.exists(f"{functions}/tick.mcfunction"):
        open(f"{functions}/tick.mcfunction", "x").close()

    if not os.path.exists(f"{tags_functions}/tick.json"):
        open(f"{tags_functions}/tick.json",
             "x").write(to_tags_json(namespace, "tick"))

    if not os.path.exists(f"{functions}/load.mcfunction"):
        open(f"{functions}/load.mcfunction", "x").close()

    if not os.path.exists(f"{tags_functions}/load.json"):
        open(f"{tags_functions}/load.json",
             "x").write(to_tags_json(namespace, "load"))

    if not os.path.exists(f"{name}/pack.mcmeta"):
        open(f"{name}/pack.mcmeta", "x").write(to_mcmeta_json(description))

    if not os.path.exists(f"{name}/project.json"):
        open(f"{name}/project.json",
             "x").write(to_project_json(name, namespace, description))
        
def publish_project(dir, auth):
    if not os.path.exists(f"{dir}/project.json"):
        print(f"Could not find project.json. Please initialise a project to publish")
        return

    with open(f"{dir}/project.json", "r") as f:
        project = json.loads(f.read())

    slug = project['slug'] or to_slug(project['name'])
    title = project['name']
    description = project['summary']
    body = project['description']
    categories = project['categories']
    license = project['license']
    version_number = project['version']

    modrinth_project: Project = Project.get(slug, auth)

    user = User.get_from_auth(auth)

    if modrinth_project:
      versions = modrinth_project.get_versions() or []

      has_version = False

      for version in versions:
          if version.version_model.version_number == version_number:
              has_version = True
              break

      modrinth_project.modify(auth=auth, title=title, description=description, body=body,  categories=categories, client_side='optional', server_side='required', license_id=license)

      if not has_version:
        publish_version(project, dir, auth)
    
    else:
      project = user.create_project(ProjectModel(
          slug, title, description, categories,
          'optional', 'required', body,
          license, 'mod'
      ))
      if project:
          print(f"Successfully published project '{title}'")
        
def publish_version(metadata, dir, auth):
    slug = metadata['slug']
    title = metadata['name']
    version_number = metadata['version']
    game_versions = metadata['game_versions']
    version_type = metadata['release_type']
    dependencies = format_dependencies(metadata["dependencies"])

    version_title = f"{title}-v{version_number}"

    zip_path = f"{dir}/{version_title}.zip"

    files_to_zip = ["data", "pack.mcmeta"]

    if os.path.exists(f"{dir}/pack.png"):
          files_to_zip.append("pack.png")

    for file in files_to_zip:
        if os.path.isdir(f"{dir}/{file}"):
          shutil.copytree(f"{dir}/{file}", file)
        else:
            shutil.copy(f"{dir}/{file}", file)
        
      
    with ZipFile(zip_path, "w") as zip_file:
        zip_file.write("data")
        zip_file.write("pack.mcmeta")
        if os.path.exists("pack.png"):
          zip_file.write("pack.png")

    for file in files_to_zip:
        if os.path.isdir(file):
          shutil.rmtree(file)
        else:
          os.remove(file)

    project = Project.get(slug, auth)

    if not project:
        print(f"Project '{title}' with slug '{slug}' was not found")
        return

    version = project.create_version(VersionModel(
        name=version_title, version_number=version_number, dependencies=dependencies,
        game_versions=game_versions, version_type=version_type, loaders=['datapack'],
        featured=True, file_parts=[zip_path]
    ), auth)

    if version:
        print(f"Successfully created version '{version_title}'")

def format_dependencies(dependencies):
    result = []

    dependency_array = list(dependencies.items())

    for dependency in dependency_array:
        slug = dependency[0]
        version_number = dependency[1]

        dependency_project = Project.get(slug)
        dependency_version = dependency_project.get_specific_version(version_number)
        formatted = {
            "project_id": dependency_project.project_model.id,
            "version_id": dependency_version.version_model.id,
            "dependency_type": "required"
        }

        result.append(formatted)
    
    return result

File: modrinth/publish.go
package modrinth

import (
	"archive/zip"
	"bufio"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"

	"github.com/gearsdatapacks/dpm/changelog"
	"github.com/gearsdatapacks/dpm/settings"
	"github.com/gearsdatapacks/dpm/types"
	"github.com/gearsdatapacks/dpm/utils"
	"github.com/gearsdatapacks/gorinth"
)

func Publish(context types.Context) {
	if context.Auth == "" {
		log.Fatal("Please specify --auth to publish a project")
	}

	project := utils.GetProjectJson()
	projectVersion, err := types.ParseVersion(project.DpmVersion)
	if err != nil {
		log.Fatal(err)
	}
	reader := bufio.NewReader(os.Stdin)
	if projectVersion.Less(settings.DPM_VERSION) {
		fmt.Print("[Warning] This project was made with an older version of DPM and may use a different project format. Consider running dpm update to update it.\nPress CTRL+C to quit or enter to continue anyway.")
		reader.ReadLine()
	}

	if projectVersion.Greater(settings.DPM_VERSION) {
		fmt.Println("[Warning] This project was made with an newer version of DPM and may use a different project format. Consider updating DPM.\nPress CTRL+C to quit or enter to continue anyway.")
		reader.ReadLine()
	}

	slug := project.Slug
	if slug == "" {
		slug = utils.ToSlug(project.Name)
	}

	modrinthProject, err := gorinth.GetProject(slug, context.Auth)

	body := modrinthProject.Body

	if filename := findFile(".", "README*", []string{}); filename != "" {
		bytes, err := os.ReadFile(filename)

		if err != nil {
			log.Fatal(err)
		}

		body = string(bytes)
	}

	var icon []byte

	if utils.Exists("pack.png") {
		icon, err = os.ReadFile("pack.png")
		if err != nil {
			log.Fatal(err)
		}
	}

	// If the project exists, modify it
	if err == nil {
		versions := modrinthProject.GetVersions()

		hasVersion := false

		for _, version := range versions {
			if version.VersionNumber == project.Version {
				hasVersion = true
				break
			}
		}

		if context.Flags.ModifyProject {
			modified := gorinth.Project{
				Title:                project.Name,
				Description:          project.Summary,
				Body:                 body,
				Categories:           project.Categories,
				AdditionalCategories: project.AdditionalCategories,
				License: &gorinth.License{
					Id: project.License,
				},
			}

			if icon != nil {
				modified.Icon = icon
			}

			err := modrinthProject.Modify(modified, context.Auth)

			if err != nil {
				log.Fatal(err)
			}

			fmt.Printf("Successfully updated project %s\n", modrinthProject.Title)
		}

		if !hasVersion && context.Flags.ModifyVersion {
			publishVersion(project, context.Auth)
		}
		return
	}
	// Otherwise, make a new one
	user := gorinth.GetUserFromAuth(context.Auth)

	toPublish := gorinth.Project{
		Slug:         slug,
		Title:        project.Name,
		Description:  project.Summary,
		Categories:   project.Categories,
		ClientSide:   "optional",
		ServerSide:   "required",
		Body:         body,
		ProjectType:  "mod",
		Status:       "draft",
		GameVersions: project.GameVersions,
		Loaders:      []string{"datapack"},
		License: &gorinth.License{
			Id: project.License,
		},
	}

	if icon != nil {
		toPublish.Icon = icon
	}

	err = user.CreateProject(toPublish)

	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Successfully created project %q\n", project.Name)

	if context.Flags.ModifyVersion {
		publishVersion(project, context.Auth)
	}
}

func findFile(dir, pattern string, excludeFiles []string) string {
	files, err := os.ReadDir(dir)
	if err != nil {
		log.Fatal(err)
	}

	for _, file := range files {
		matched, err := filepath.Match(pattern, file.Name())
		if err != nil {
			log.Fatal(err)
		}

		if !matched {
			continue
		}

		matchedExclude := false
		for _, exclude := range excludeFiles {
			matched, err := filepath.Match(exclude, file.Name())
			if err != nil {
				log.Fatal(err)
			}
			if matched {
				matchedExclude = true
			}
		}

		if !matchedExclude {
			return file.Name()
		}
	}

	return ""
}

func publishVersion(metadata types.Project, auth string) {
	title := metadata.Name
	versionNumber := metadata.Version
	dependencies := formatDependencies(metadata.Dependencies, "required", auth)
	optionalDependencies := formatDependencies(metadata.OptionalDependencies, "optional", auth)
	dependencies = append(dependencies, optionalDependencies...)

	versionTitle := fmt.Sprintf("%s v%s", title, versionNumber)

	zipPath := versionTitle + ".zip"

	filesToZip := []string{"data", "pack.mcmeta"}

	includedFiles := []string{
		"pack.png",
		"README*",
		"CHANGELOG*",
		"LICENSE*",
	}
	excludedFiles := []string{}

	var changelogText string

	if filename := findFile(".", "CHANGELOG*", []string{}); filename != "" {
		version, err := types.ParseVersion(metadata.Version)
		if err == nil {
			changelogText = changelog.ParseChangelog(filename, *version)
		}
	}

	config, hasConfig := utils.GetConfig()

	if hasConfig {
		includedFiles = append(includedFiles, config.IncludeFiles...)
		excludedFiles = config.ExcludeFiles
	}

	for _, file := range includedFiles {
		if filename := findFile(".", file, excludedFiles); filename != "" {
			filesToZip = append(filesToZip, filename)
		}
	}

	zipFiles(zipPath, filesToZip)

	project, err := gorinth.GetProject(metadata.Slug, auth)

	if err != nil {
		log.Fatal(err)
	}

	version := gorinth.Version{
		Name:          versionTitle,
		VersionNumber: versionNumber,
		Dependencies:  dependencies,
		GameVersions:  metadata.GameVersions,
		VersionType:   metadata.ReleaseType,
		Loaders:       []string{"datapack"},
		FileParts:     []string{zipPath},
		Changelog:     changelogText,
	}

	err = project.CreateVersion(version, auth)

	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Successfully created version %q\n", versionTitle)
}

func formatDependencies(dependencies map[string]string, depType string, auth string) []gorinth.Dependency {
	result := []gorinth.Dependency{}

	for slug, versionNumber := range dependencies {
		dependency, err := gorinth.GetProject(slug, auth)
		if err != nil {
			log.Fatal(err)
		}
		version := dependency.GetSpecificVersion(versionNumber)

		formatted := gorinth.Dependency{
			ProjectId:      dependency.Id,
			VersionId:      version.Id,
			DependencyType: depType,
		}

		result = append(result, formatted)
	}

	return result
}

func zipFiles(filename string, files []string) {
	archive, err := os.Create(filename)
	if err != nil {
		log.Fatal(err)
	}
	defer archive.Close()

	zipWriter := zip.NewWriter(archive)
	defer zipWriter.Close()

	for _, fileToZip := range files {
		if utils.IsDir(fileToZip) {
			zipDir(zipWriter, fileToZip)
		} else {
			zipFile(zipWriter, fileToZip)
		}
	}
}

func zipFile(writer *zip.Writer, fileToZip string) {
	file, err := os.Open(fileToZip)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	zipped, err := writer.Create(fileToZip)
	if err != nil {
		log.Fatal(err)
	}

	_, err = io.Copy(zipped, file)
	if err != nil {
		log.Fatal(err)
	}
}

func zipDir(writer *zip.Writer, dir string) {
	files, err := os.ReadDir(dir)
	if err != nil {
		log.Fatal(err)
	}

	for _, file := range files {
		path := fmt.Sprintf("%s/%s", dir, file.Name())
		if file.IsDir() {
			zipDir(writer, path)
		} else {
			zipFile(writer, path)
		}
	}
}
